cupyx/scipy/ndimage/_interp_kernels.py

```diff
@@ -0,0 +1,400 @@
+            ops.append("""
+            for (int s_{j} = 0; s_{j} < n_{j}; s_{j}++)
+                {{
+                    W w_{j};
+                    {int_t} ic_{j};
+                    if (s_{j} == 0)
+                    {{
+                        w_{j} = (W)cc_{j} - c_{j};
+                        ic_{j} = cf_bounded_{j} * sx_{j};
+                    }} else
+                    {{
+                        w_{j} = c_{j} - (W)cf_{j};
+                        ic_{j} = cc_bounded_{j} * sx_{j};
+                    }}""".format(int_t=int_t, j=j))
+
+        _weight = ' * '.join(['w_{j}'.format(j=j) for j in range(ndim)])
+        _coord_idx = ' + '.join(['ic_{j}'.format(j=j) for j in range(ndim)])
+        ops.append("""
+        X val = x[{coord_idx}];
+        out += val * ({weight});
+        """.format(coord_idx=_coord_idx, weight=_weight))
+        ops.append('}' * ndim)
+
+    if mode == 'constant':
+        ops.append('}')
+
+    if integer_output:
+        ops.append('y = (Y)rint((double)out);')
+    else:
+        ops.append('y = (Y)out;')
+    operation = '\n'.join(ops)
+
+    name = 'interpolate_{}_order{}_{}_x{}_y{}'.format(
+        name,
+        order,
+        mode,
+        '_'.join(['{}'.format(j) for j in xshape]),
+        '_'.join(['{}'.format(j) for j in yshape]),
+    )
+    return operation, name
+
+
+@cupy.util.memoize()
+def _get_map_kernel(xshape, mode, cval=0.0, order=1, integer_output=False):
+    in_params = 'raw X x, raw W coords'
+    out_params = 'Y y'
+    operation, name = _generate_interp_custom(
+        coord_func=_get_coord_map,
+        xshape=xshape,
+        yshape=xshape,
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='shift',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
+def _get_shift_kernel(xshape, yshape, mode, cval=0.0, order=1,
+                      integer_output=False):
+    in_params = 'raw X x, raw W shift'
+    out_params = 'Y y'
+    operation, name = _generate_interp_custom(
+        coord_func=_get_coord_shift,
+        xshape=xshape,
+        yshape=yshape,
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='shift',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
+def _get_zoom_shift_kernel(xshape, yshape, mode, cval=0.0, order=1,
+                           integer_output=False):
+    in_params = 'raw X x, raw W shift, raw W zoom'
+    out_params = 'Y y'
+    operation, name = _generate_interp_custom(
+        coord_func=_get_coord_zoom_and_shift,
+        xshape=xshape,
+        yshape=yshape,
+        mode=mode,
+        cval=cval,
+        order=order,
+        name='zoom_shift',
+        integer_output=integer_output,
+    )
+    return cupy.ElementwiseKernel(in_params, out_params, operation, name)
+
+
+@cupy.util.memoize()
```

Comment:
```suggestion
@cupy.util.memoize(for_each_device=True)
```

Reply:
Thanks!

I didn't want to make separate commits for all of these so I did one commit manually and added
`Co-authored-by: Leo Fang <leofang@bnl.gov>` there

Reply:
Thanks, Gregory. No worries about the coauthor thing. I just thought there's a batch commit button and having all needed changes explicitly laid out might be easier for you.